src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ValuesSources/WeaklyCachedRecoverableValueSource.cs

```diff
@@ -6,26 +6,31 @@
 using System.Diagnostics.CodeAnalysis;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Host
 {
     /// <summary>
-    /// This class is a <see cref="ValueSource{T}"/> that holds onto a value weakly, 
-    /// but can save its value and recover it on demand if needed.
-    /// 
-    /// The initial value comes from the <see cref="ValueSource{T}"/> specified in the constructor.
-    /// Derived types implement SaveAsync and RecoverAsync.
+    /// This class is a <see cref="ValueSource{T}"/> that holds onto a value weakly, but can save its value and recover
+    /// it on demand if needed.  The value is initially strongly held, until the first time that <see cref="GetValue"/>
+    /// or <see cref="GetValueAsync"/> is called.  At that point, it will be dumped to secondary storage, and retrieved
+    /// and weakly held from that point on in the future.
     /// </summary>
     internal abstract class WeaklyCachedRecoverableValueSource<T> : ValueSource<T> where T : class
     {
+        // enforce saving in a queue so save's don't overload the thread pool.
+        private static Task s_latestTask = Task.CompletedTask;
+        private static readonly NonReentrantLock s_taskGuard = new();
```

Comment:
these fields just moved up, they're not new.

Reply:
Not just moved up -- you're also bringing this code into the 20th century with some doc comments!